File: packages/core-py/src/ebb_ai/providers/ollama.py
"""Ollama adapter.

Talks to a local Ollama server over HTTP (default http://localhost:11434,
override with ``OLLAMA_HOST``). No API key — Ollama runs on the machine. Uses
``httpx`` directly; no vendor SDK dependency.

Endpoint used:
  - ``POST /api/chat`` (stream:false) — sync dispatch, returns the assistant
    message plus ``prompt_eval_count`` / ``eval_count`` token counts.

Batch: UNSUPPORTED. Ollama is local inference with no batch API — there is no
cost/SLA batch tier to route through. This adapter inherits the base
``dispatch_batch`` / ``retrieve_batch`` stubs (both raise); the scheduler
keeps Ollama tasks on the sync path.

Carbon accounting is unchanged: energy coefficients for local models
(``llama-*``, ``mistral-*``, ``mixtral-*``) already live in the SSOT energy
table, and the receipt applies the caller's grid intensity exactly as it does
for hosted providers. There is no special-cased "local" carbon logic here.
"""

from __future__ import annotations

import os
import re
from typing import Any

import httpx

from .base import DispatchOptions, DispatchResult, ProviderAdapter

_DEFAULT_HOST = "http://localhost:11434"


def _normalize_host(raw: str | None) -> str:
    """Normalize a host into a scheme-qualified base URL with no trailing slash."""
    host = (raw or _DEFAULT_HOST).strip() or _DEFAULT_HOST
    if not re.match(r"^https?://", host, re.IGNORECASE):
        host = f"http://{host}"
    return host.rstrip("/")


class OllamaAdapter(ProviderAdapter):
    """Adapter for a local Ollama server.

    Parameters
    ----------
    host:
        Base host URL. If omitted, read from ``OLLAMA_HOST``, else
        ``http://localhost:11434``. A bare ``host:port`` is accepted and
        prefixed with ``http://``.
    client:
        Optional pre-built ``httpx.AsyncClient`` (or any object exposing an
        awaitable ``post(url, *, json)`` returning a response with
        ``status_code`` / ``json()`` / ``text``). Injected in tests to avoid
        live calls. When omitted, a short-lived client is created per call.
    timeout_s:
        Per-request timeout when constructing the default client. Local
        inference can be slow, so this is generous.
    """

    name = "ollama"

    def __init__(
        self,
        *,
        host: str | None = None,
        client: Any | None = None,
        timeout_s: float = 300.0,
    ) -> None:
        self._host = _normalize_host(host if host is not None else os.environ.get("OLLAMA_HOST"))
        self._client = client
        self._timeout_s = timeout_s

    @property
    def ready(self) -> bool:
        """Always True: Ollama is local + keyless. Reachability of the local
        server surfaces at dispatch time (a connection-refused there), not
        here."""
        return True

    async def dispatch(
        self,
        model: str,
        prompt: str,
        options: DispatchOptions | None = None,
    ) -> DispatchResult:
        opts = options or DispatchOptions()

        ollama_options: dict[str, Any] = {}
        temperature = opts.extra.get("temperature")
        if temperature is not None:
            ollama_options["temperature"] = temperature
        if opts.max_tokens is not None:
            ollama_options["num_predict"] = opts.max_tokens

        messages: list[dict[str, str]] = []
        if opts.system is not None:
            messages.append({"role": "system", "content": opts.system})
        messages.append({"role": "user", "content": prompt})

        body: dict[str, Any] = {"model": model, "messages": messages, "stream": False}
        if ollama_options:
            body["options"] = ollama_options

        url = f"{self._host}/api/chat"
        try:
            if self._client is not None:
                res = await self._client.post(url, json=body)
            else:
                async with httpx.AsyncClient(
                    timeout=httpx.Timeout(self._timeout_s)
                ) as client:
                    res = await client.post(url, json=body)
        except httpx.HTTPError as err:
            # Connection refused / DNS / TLS — the local server is unreachable.
            raise RuntimeError(
                f"OllamaAdapter: could not reach Ollama at {self._host} ({err}). "
                "Is `ollama serve` running? Set OLLAMA_HOST to override the address."
            ) from err

        if res.status_code != httpx.codes.OK:
            raise RuntimeError(f"Ollama API {res.status_code}: {res.text[:300]}")

        data = res.json()
        message = data.get("message") or {}
        return DispatchResult(
            text=message.get("content") or "",
            model=data.get("model") or model,
            provider=self.name,
            raw=data,
            input_tokens=data.get("prompt_eval_count"),
            output_tokens=data.get("eval_count"),
        )


__all__ = ["OllamaAdapter"]

File: packages/core-ts/src/providers/ollama.ts
/**
 * Ollama provider adapter.
 *
 * Talks to a local Ollama server over HTTP (default
 * http://localhost:11434, override with `OLLAMA_HOST`). No API key — Ollama
 * runs on the machine. Uses `fetch` directly; no vendor SDK dependency.
 *
 * Endpoint used:
 *   - POST /api/chat (stream:false) — sync dispatch, returns the assistant
 *     message plus `prompt_eval_count` / `eval_count` token counts.
 *
 * Batch: UNSUPPORTED. Ollama is local inference with no batch API — there is
 * no cost/SLA batch tier to route through. `dispatchBatch` is omitted; the
 * scheduler keeps Ollama tasks on the sync path.
 *
 * Carbon accounting is unchanged: energy coefficients for local models
 * (llama-*, mistral-*, mixtral-*) already live in the SSOT energy table, and
 * the receipt math applies the caller's grid intensity exactly as it does for
 * hosted providers. There is no special-cased "local" carbon logic here.
 */

import type {
  DispatchOptions,
  DispatchResult,
  ProviderAdapter,
} from "./base.js";

/** Minimal shape of the `fetch` we depend on (injectable for tests). */
type FetchLike = (
  url: string,
  init: {
    method: string;
    headers: Record<string, string>;
    body: string;
  },
) => Promise<{
  ok: boolean;
  status: number;
  text(): Promise<string>;
  json(): Promise<unknown>;
}>;

export interface OllamaAdapterOptions {
  /**
   * Base host URL. Defaults to `process.env.OLLAMA_HOST`, else
   * `http://localhost:11434`. A bare `host:port` is accepted and prefixed
   * with `http://`.
   */
  host?: string;
  /** Inject a `fetch` implementation. Defaults to the global `fetch`. */
  fetchImpl?: FetchLike;
}

const DEFAULT_HOST = "http://localhost:11434";

/** Normalize a host string into a scheme-qualified base URL with no trailing slash. */
function normalizeHost(raw: string | undefined): string {
  const h = (raw ?? DEFAULT_HOST).trim() || DEFAULT_HOST;
  const withScheme = /^https?:\/\//i.test(h) ? h : `http://${h}`;
  return withScheme.replace(/\/$/, "");
}

export class OllamaAdapter implements ProviderAdapter {
  readonly provider = "ollama";
  private readonly host: string;
  private readonly fetchImpl: FetchLike | undefined;

  constructor(opts: OllamaAdapterOptions = {}) {
    this.host = normalizeHost(opts.host ?? process.env.OLLAMA_HOST);
    this.fetchImpl = opts.fetchImpl;
  }

  /**
   * Always "ready": Ollama needs no credential. Whether the local server is
   * actually reachable is only known at dispatch time (a connection-refused
   * surfaces there, not here).
   */
  get ready(): boolean {
    return true;
  }

  async dispatch(
    model: string,
    prompt: string,
    options: DispatchOptions = {},
  ): Promise<DispatchResult> {
    const fetchImpl = this.fetchImpl ?? (globalThis.fetch as FetchLike | undefined);
    if (typeof fetchImpl !== "function") {
      throw new Error(
        "OllamaAdapter: no fetch implementation available. Pass { fetchImpl } or run on a runtime with global fetch.",
      );
    }

    const ollamaOptions: Record<string, unknown> = {};
    if (options.temperature !== undefined) {
      ollamaOptions.temperature = options.temperature;
    }
    if (options.maxTokens !== undefined) {
      ollamaOptions.num_predict = options.maxTokens;
    }

    const messages = options.system
      ? [
          { role: "system", content: options.system },
          { role: "user", content: prompt },
        ]
      : [{ role: "user", content: prompt }];

    const body: Record<string, unknown> = {
      model,
      messages,
      stream: false,
    };
    if (Object.keys(ollamaOptions).length > 0) {
      body.options = ollamaOptions;
    }

    const url = `${this.host}/api/chat`;
    let res: Awaited<ReturnType<FetchLike>>;
    try {
      res = await fetchImpl(url, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(body),
      });
    } catch (err) {
      // Connection refused / DNS / TLS — the local server is not reachable.
      const detail = err instanceof Error ? err.message : String(err);
      throw new Error(
        `OllamaAdapter: could not reach Ollama at ${this.host} (${detail}). ` +
          `Is \`ollama serve\` running? Set OLLAMA_HOST to override the address.`,
      );
    }
    if (!res.ok) {
      let detail = "";
      try {
        detail = (await res.text()).slice(0, 300);
      } catch {
        detail = "(no response body)";
      }
      throw new Error(`Ollama API ${res.status}: ${detail}`);
    }
    const data = (await res.json()) as {
      message?: { content?: string };
      prompt_eval_count?: number;
      eval_count?: number;
      model?: string;
    };
    const inputTokens = data.prompt_eval_count;
    const outputTokens = data.eval_count;
    return {
      text: data.message?.content ?? "",
      usage: {
        inputTokens,
        outputTokens,
        totalTokens:
          inputTokens !== undefined && outputTokens !== undefined
            ? inputTokens + outputTokens
            : undefined,
      },
      model: data.model ?? model,
      provider: this.provider,
      raw: data,
    };
  }

  // No dispatchBatch / retrieveBatch — Ollama has no batch API.
}
